packages/contracts/src/contracts/current/protocol/Exchange/MixinSignatureValidator.sol

```diff
@@ -0,0 +1,117 @@
+/*
+
+  Copyright 2017 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.4.19;
+
+import "./mixins/MSignatureValidator.sol";
+import "./ISigner.sol";
+
+/// @dev Provides MSignatureValidator
+contract MixinSignatureValidator is
+    MSignatureValidator
+{
+    enum SignatureType {
+        Invalid,
+        Caller,
+        Ecrecover,
+        EIP712,
+        Contract
+    }
+  
+    function isValidSignature(
+        bytes32 hash,
+        address signer,
+        bytes signature)
+        public view
+        returns (bool isValid)
+    {
+        // TODO: Domain separation: make hash depend on role. (Taker sig should not be valid as maker sig, etc.)
+        
+        require(signature.length >= 1);
+        SignatureType stype = SignatureType(uint8(signature[0]));
+        
+        // Variables are not scoped in Solidity
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        
+        // Zero is always an invalid signature
+        if (stype == SignatureType.Invalid) {
+            require(signature.length == 1);
+            isValid = false;
+            return;
+        
+        // Implicitely signed by caller
+        } else if (stype == SignatureType.Caller) {
+            require(signature.length == 1);
```

Comment:
Can you give an example of how this would be used? It seems like `msg.sender` will always be `order.maker` in this case, which seems a bit strange to me.